recipes/lora_finetune_single_device.py

```diff
@@ -507,6 +507,14 @@ def save_checkpoint(self, epoch: int) -> None:
         # Move to CPU to avoid a copy on GPU
         state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
 
+        # Construct the adapter weights
+        # Do this using the state_dict to avoid running upcast and H2D in state_dict post hook twice
+        # Must be before get_merged_lora_ckpt because get_merged_lora_ckpt will remove lora keys
+        adapter_key_filter = lambda x: x in self.adapter_params
+        adapter_state_dict = {
+            k: v for k, v in state_dict.items() if adapter_key_filter(k)
+        }
```

Comment:
Ideally we want to only run state_dict post hooks once, so we should reuse the state_dict

this does change the semantic slightly though -- before `adapter_0.pt` contained weights tagged with CUDA, but now it contains weights tagged with CPU.

 Not sure whether the old behavior was intended/whether this change is ok (but no CI seems to fail :D) Also when loading, we [`map_location="cpu"` regardless](https://github.com/pytorch/torchtune/blob/81bc3daaf15e84c0a40d813f6eed996595eb422a/torchtune/utils/_checkpointing/_checkpointer_utils.py#L110)

Reply:
Sorry I missed this comment before. But yeah I think this change makes sense, I don't think there's any reason we need to require CUDA weights. And as you point out since we load on CPU when resuming it was probably never really an issue. Plus not re-running the state dict post hooks is a nice bonus.